Mage.Sets/src/mage/cards/c/CavernOfSouls.java

```diff
@@ -186,8 +185,10 @@ public boolean checksEventType(GameEvent event, Game game) {
 
     @Override
     public boolean applies(GameEvent event, Ability source, Game game) {
+
         CavernOfSoulsWatcher watcher = game.getState().getWatcher(CavernOfSoulsWatcher.class, source.getSourceId());
         Spell spell = game.getStack().getSpell(event.getTargetId());
-        return spell != null && watcher != null && watcher.spellCantBeCountered(spell.getId());
+        Integer currentZCC = game.getObject(event.getTargetId()).getZoneChangeCounter(game);
+        return spell != null && watcher != null && watcher.spellCantBeCountered(spell.getId(), currentZCC);
```

Comment:
There are a special `MageObjectReference` object to help with all ZCC code. Store card ref in Set<MageObjectReference> and call getCard from that ref to get actual card or null (if zcc changed).

Example:
![shot_200612_213801](https://user-images.githubusercontent.com/8344157/84531463-b745ee80-acf5-11ea-9e83-23f62f45b537.png)

Reply:
Good to know! I've wondered before what those references were doing